lib/Onyx.d.ts

```diff
@@ -0,0 +1,294 @@
+import {Component} from 'react';
+import {PartialDeep} from 'type-fest';
+import {CustomTypeOptions} from '.';
+import * as Logger from './Logger';
+
+type MergeBy<T, K> = Omit<T, keyof K> & K;
+type DeepRecord<K extends string | number | symbol, T> = {[key: string]: T | DeepRecord<K, T>};
+type DeepKeyOf<T> = T extends object
+    ? {
+          [K in keyof T & (string | number)]: T[K] extends Record<string, unknown> ? `${K}.${DeepKeyOf<T[K]>}` | K : K;
+      }[keyof T & (string | number)]
+    : T;
+
+type TypeOptions = MergeBy<
+    {
+        keys: string;
+        values: Record<string, unknown>;
+    },
+    CustomTypeOptions
+>;
+
+type Key = TypeOptions['keys'];
+type Value = TypeOptions['values'];
+
+type KeyValueMap = {
+    [K in Key]: Value[K];
+};
+
+type BaseConnectOptions<K extends Key> = {
+    key: K;
+    statePropertyName?: string;
+    withOnyxInstance?: Component;
+    initWithStoredValues?: boolean;
+    selector?: Value[K] extends object ? ((value: Value[K] | null) => PartialDeep<Value[K]> | null) | DeepKeyOf<Value[K]> : never;
+};
+
+declare type ConnectOptions<K extends Key> = BaseConnectOptions<K> &
+    (
+        | {
+              callback?: (value: Record<string, Value[K]> | null, key?: K) => void;
+              waitForCollectionCallback: true;
+          }
+        | {
+              callback?: (value: Value[K] | null, key?: K) => void;
+              waitForCollectionCallback?: false;
+          }
+    );
+
+declare type MergeCollection<K extends Key, Map, Value> = {
+    [MapK in keyof Map]: MapK extends `${K}${string}`
+        ? MapK extends `${K}`
+            ? never // forbids empty id
+            : Value
+        : never;
+};
+
+declare type UpdateOperation<K extends Key> =
+    | {
+          onyxMethod: 'set';
+          key: K;
+          value: Value[K];
+      }
+    | {
+          onyxMethod: 'merge';
+          key: K;
+          value: PartialDeep<Value[K]>;
+      }
+    | {
+          onyxMethod: 'mergeCollection'; // TODO: Finish this.
```

Comment:
@fabioh8010 I wasn't able to type this unfortunately. I tried several different approaches where I tried to use `MergeCollection` but unsuccessfully. Here is what kind of works but isn't fully type safe:
```ts
   | {
          onyxMethod: 'mergeCollection';
          key: K;
          value: Record<`${K}${string}`, PartialDeep<Value[K]>>;
      };
```